### PR TITLE
Issue 1766 Modified mustache files for Go to support nullable in the …

### PR DESCRIPTION
**Pull request by alex-korobko**
…spec v3.0+; Updated model files running .sh scripts for Go.

### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [x] Filed the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`~~, `3.4.x`, `4.0.x`~~. Default: `master`.
- [x] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR

Updated mustache files to enable the `nullable` support (https://github.com/OpenAPITools/openapi-generator/issues/1766)
Ran all ./bin/go*.sh files to update the models; 
Not sure though how to test the changes though/ The integration test (as it recommended in the guide) classes for the Golang are using specs from the ver 2.0, so it seems like there is no reason to use them. Should we try to make new classes for that? Do you have any guide for that? 

@antihax  @bvwells @grokify @kemokemo

### CODE REVIEW

File: modules/openapi-generator/src/main/resources/go-gin-server/model.mustache [hunk 1]
```diff
@@ -19,6 +19,6 @@ const (
 type {{classname}} struct {
 {{#vars}}{{#description}}
 	// {{{description}}}{{/description}}
-	{{name}} {{^isEnum}}{{^isPrimitiveType}}{{^isContainer}}{{^isDateTime}}*{{/isDateTime}}{{/isContainer}}{{/isPrimitiveType}}{{/isEnum}}{{{dataType}}} `json:"{{baseName}}{{^required}},omitempty{{/required}}"{{#vendorExtensions.x-go-custom-tag}} {{{.}}}{{/vendorExtensions.x-go-custom-tag}}`
```
Comment (alex-korobko) [inline modules/openapi-generator/src/main/resources/go-gin-server/model.mustache]: This is the change that I'd like to discuss. In a previous version relative model objects were generated as classes. As we have `nullable` support now, I don't think we need to keep this implicit behavior. I'd rather expect to have the pointer defined as it should be -  using the `nullable` property.

File: modules/openapi-generator/src/main/resources/go-server/model.mustache [hunk 2]
```diff
@@ -19,6 +19,6 @@ const (
 type {{classname}} struct {
 {{#vars}}{{#description}}
 	// {{{description}}}{{/description}}
-	{{name}} {{^isEnum}}{{^isPrimitiveType}}{{^isContainer}}{{^isDateTime}}*{{/isDateTime}}{{/isContainer}}{{/isPrimitiveType}}{{/isEnum}}{{{dataType}}} `json:"{{baseName}}{{^required}},omitempty{{/required}}"{{#vendorExtensions.x-go-custom-tag}} {{{.}}}{{/vendorExtensions.x-go-custom-tag}}`
```
Comment (alex-korobko) [inline modules/openapi-generator/src/main/resources/go-server/model.mustache]: This is the change that I'd like to discuss. In a previous version relative model objects were generated as classes. As we have `nullable` support now, I don't think we need to keep this implicit behavior. I'd rather expect to have the pointer defined as it should be -  using the `nullable` property.

File: modules/openapi-generator/src/main/resources/go-gin-server/model.mustache (hunk 1 above)
Comment (kemokemo) [reply modules/openapi-generator/src/main/resources/go-gin-server/model.mustache]: Thank you for your PR! 😃 
I agree with you.

**Comment by kemokemo** [comment]
I have made a brief review.
There is something I want to check one point so please give me some time.

**Comment by ackintosh** [comment]
@alex-korobko Thanks for the PR! 👍 

---

@OpenAPITools/generator-core-team 

> Updated mustache files to enable the nullable support (#1766)
> Ran all ./bin/go*.sh files to update the models;
> Not sure though how to test the changes though

I think we need to add `nullable` to ["3_0/petstore-with-fake-endpoints-models-for-testing.yaml"](https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/test/resources/3_0/petstore-with-fake-endpoints-models-for-testing.yaml) to make sure that the output is generated as we expected. do you have any concerns?

**Comment by alex-korobko** [comment]
> @alex-korobko Thanks for the PR! 👍
> 
> TO: @OpenAPITools/generator-core-team
> 
> [#1869 (comment)](https://github.com/OpenAPITools/openapi-generator/pull/1869#issue-243488323)
> 
> > Updated mustache files to enable the nullable support (#1766)
> > Ran all ./bin/go*.sh files to update the models;
> > Not sure though how to test the changes though
> 
> I think we need to add `nullable` to ["3_0/petstore-with-fake-endpoints-models-for-testing.yaml"](https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/test/resources/3_0/petstore-with-fake-endpoints-models-for-testing.yaml) to make sure that the output is generated as we expected. do you have any concerns?

Not sure if it could help, the ["3_0/petstore-with-fake-endpoints-models-for-testing.yaml"] file is not used in go test scripts yet, so we should probably need more changes in order to get `nullable` testable for Golang model generation.

**Comment by ackintosh** [comment]
Sorry I didn't explain it enough. 💦
I think we need to add a shell script under `./bin/*openapi3*` like [bin/openapi3/php-petstore.sh](https://github.com/OpenAPITools/openapi-generator/blob/master/bin/openapi3/php-petstore.sh) before adding `nullable` to the yaml file. 
(The script uses `petstore-with-fake-endpoints-models-for-testing.yaml`)

**Comment by alex-korobko** [comment]
> Sorry I didn't explain it enough. 💦
> I think we need to add a shell script under `./bin/*openapi3*` like [bin/openapi3/php-petstore.sh](https://github.com/OpenAPITools/openapi-generator/blob/master/bin/openapi3/php-petstore.sh) before adding `nullable` to the yaml file.
> (The script uses `petstore-with-fake-endpoints-models-for-testing.yaml`)

Got it, thanks! So who should create that Shell script to test v3 spec features for Go? Did you recommend me to it to make the script in this PR or guys from the team are adding such scripts using some other internal workflow?

**Comment by kemokemo** [comment]
Sorry for the late review. 
I have completed checking for the code you had already committed. They looks good. 👍 
There is still content to discuss, but I will report it. 👀

**Comment by kemokemo** [comment]
To: @alex-korobko  @ackintosh 

Hi!

I had added the nullable property to the  `petstore-with-fake-endpoints-models-for-testing.yaml` and created `go-petstore.sh`.
I could not push to [alex-korobko/openapi-generator](https://github.com/alex-korobko/openapi-generator) repository so I saved it to my gist. Sorry.. 🙇 

- [`modules/openapi-generator/src/test/resources/3_0/petstore-with-fake-endpoints-models-for-testing.yaml`](https://gist.github.com/kemokemo/3625abb448b61ba0d66d3e72a477b0ab#file-petstore-with-fake-endpoints-models-for-testing-yaml)
- [`bin/openapi3/go-petstore.sh`](https://gist.github.com/kemokemo/3625abb448b61ba0d66d3e72a477b0ab#file-go-petstore-sh)

Can you confirm the content? Also, if it seems good, please apply it to this repository.
Then I want you to run the script and update the sample.

**Comment by ackintosh** [comment]
@kemokemo Thanks for your good job! I'm working on that applying your changes to this PR. 😉

**Comment by ackintosh** [comment]
[The updates](https://github.com/OpenAPITools/openapi-generator/pull/1869#issuecomment-459950611) from @kemokemo has been pushed to this PR. Please take a look if [the sample](https://github.com/OpenAPITools/openapi-generator/pull/1869/commits/376e1bacac38bb51138665de638b01ab0cccfb75) is generated as you guys expected. @kemokemo @alex-korobko

**Comment by alex-korobko** [comment]
Thanks a lot for your help! Let me check the updates.

**Comment by ackintosh** [comment]
FYI: [The updates before](https://github.com/OpenAPITools/openapi-generator/pull/1869#issuecomment-460579060) has generated only "go client", 
so in order to generate "go gin server" and "go server" samples I've added [`bin/openapi3/go-gin-petstore-server.sh`](https://github.com/OpenAPITools/openapi-generator/pull/1869/commits/20677e8802a4118ee2c350a41693067efb2d1475) and updated [`bin/openapi3/go-petstore-server.sh`](https://github.com/OpenAPITools/openapi-generator/pull/1869/commits/90e4bb2a1dbdb670959a2c9deec292381de6e10b).

**Comment by ackintosh** [comment]
The samples added by this PR:


- client
  - https://github.com/OpenAPITools/openapi-generator/pull/1869/commits/376e1bacac38bb51138665de638b01ab0cccfb75
- server
  - https://github.com/OpenAPITools/openapi-generator/pull/1869/commits/58d2b1454761251926175f9806d893fa4d5831a9
- gin server
  - https://github.com/OpenAPITools/openapi-generator/pull/1869/commits/ed1dde6940a7688852e9dd53787041cc6594ce59

It looks good for me as the parameter defined as `nullable` is generated as pointer like below.

> NullableMessage *string \`json:"NullableMessage,omitempty"\`

**Comment by kemokemo** [comment]
@ackintosh  Thank you so much for your merging! 😄 
I had checked the sample code generated using `petstore-with-fake-endpoints-models-for-testing.yaml` and found a potential problem with `go-server` and `go-gin-server`.

## issue

The `import` statement is duplicated.

* [go-api-server's sample](https://github.com/alex-korobko/openapi-generator/blob/58d2b1454761251926175f9806d893fa4d5831a9/samples/openapi3/server/petstore/go-api-server/go/model_format_test_.go#L12-L18)
* [go-gin-api-server's sample](https://github.com/alex-korobko/openapi-generator/blob/58d2b1454761251926175f9806d893fa4d5831a9/samples/openapi3/server/petstore/go-gin-api-server/go/model_format_test_.go#L12-L18)

## plan to resolve above

```diff
 {{>partial_header}}
 package {{packageName}}
 {{#models}}{{#imports}}
-import ({{/imports}}{{#imports}}
-	"{{import}}"{{/imports}}{{#imports}}
-)
+{{#-first}}import ({{/-first}}
+    "{{import}}"
+{{#-last}}){{/-last}}
 {{/imports}}{{#model}}{{#isEnum}}{{#description}}// {{{classname}}} : {{{description}}}{{/description}}
 type {{{name}}} {{^format}}{{dataType}}{{/format}}{{#format}}{{{format}}}{{/format}}
```

I am sorry many times.
Could @ackintosh or @alex-korobko please apply the above contents to the following file? 🙇 

* [go-server/model.mustache](https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/go-server/model.mustache)
* [go-gin-server/model.mustache](https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/go-gin-server/model.mustache)

**Comment by ackintosh** [comment]
I'm working on [the issue](https://github.com/OpenAPITools/openapi-generator/pull/1869#issuecomment-462104862). 😉

**Comment by ackintosh** [comment]
I've updated the templates, please have a look when you have time. ✨

**Comment by kemokemo** [comment]
@ackintosh , thanks a lot for your support! 😉 
It looks very nice. My review has been completed.

Everyone, please point out if you notice any problems.
Thank you!

File: samples/openapi3/client/petstore/go/go-petstore/model_health_check_result.go [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+ *
+ * API version: 1.0.0
+ * Generated by: OpenAPI Generator (https://openapi-generator.tech)
+ */
+
+package petstore
+
+// Just a string to inform instance is up and running. Make it nullable in hope to get it as pointer in generated model.
+type HealthCheckResult struct {
+	NullableMessage *string `json:"NullableMessage,omitempty"`
```
Comment (alex-korobko) [inline samples/openapi3/client/petstore/go/go-petstore/model_health_check_result.go]: Looks good to me, thanks a lot!

**Comment by ackintosh** [comment]
The PR has been merged into master. Thanks for your contribution! @alex-korobko @kemokemo 👍 👍 👍

**Pull request merged**